File: src/security.ts
import argon2 from "argon2";
import { createHmac, randomBytes } from "node:crypto";

export const MAX_PASSWORD_LENGTH = 128;
export const MAX_EMAIL_LENGTH = 254;

export function normalizeEmail(value: string): string {
  return value.trim().toLowerCase().slice(0, MAX_EMAIL_LENGTH);
}

export function validatePassword(password: string): string | null {
  if (password.length < 8) {
    return "Password must be at least 8 characters.";
  }
  if (password.length > MAX_PASSWORD_LENGTH) {
    return `Password must be at most ${MAX_PASSWORD_LENGTH} characters.`;
  }
  if (!/[a-zA-Z]/.test(password)) {
    return "Password must include at least one letter.";
  }
  if (!/\d/.test(password)) {
    return "Password must include at least one number.";
  }
  return null;
}

export async function hashPassword(password: string): Promise<string> {
  return argon2.hash(password);
}

export async function verifyPassword(passwordHash: string, password: string): Promise<boolean> {
  try {
    return await argon2.verify(passwordHash, password);
  } catch {
    return false;
  }
}

export function generateSessionToken(): string {
  return randomBytes(32).toString("base64url");
}

export function generateOneTimeToken(): string {
  return randomBytes(32).toString("base64url");
}

export function hashSessionToken(token: string, secret: string): string {
  return createHmac("sha256", secret).update(token).digest("hex");
}

File: packages/fire_dashboard/auth_api/shared/security.py
from __future__ import annotations

import hashlib
import hmac
import secrets
from typing import Any

def _password_hasher() -> Any:
    from argon2 import PasswordHasher

    return PasswordHasher()


def normalize_email(value: str) -> str:
    return value.strip().lower()


def validate_password(password: str) -> str | None:
    if len(password) < 8:
        return "Password must be at least 8 characters."
    if not any(ch.isalpha() for ch in password):
        return "Password must include at least one letter."
    if not any(ch.isdigit() for ch in password):
        return "Password must include at least one number."
    return None


def hash_password(password: str) -> str:
    return _password_hasher().hash(password)


def verify_password(password_hash: str, password: str) -> bool:
    from argon2.exceptions import VerifyMismatchError

    try:
        return _password_hasher().verify(password_hash, password)
    except VerifyMismatchError:
        return False


def generate_session_token() -> str:
    return secrets.token_urlsafe(32)


def generate_one_time_token() -> str:
    return secrets.token_urlsafe(32)


def hash_session_token(token: str, secret: str) -> str:
    return hmac.new(secret.encode("utf-8"), token.encode("utf-8"), hashlib.sha256).hexdigest()


def safe_equal_string(left: str, right: str) -> bool:
    left_bytes = left.encode("utf-8")
    right_bytes = right.encode("utf-8")
    if len(left_bytes) != len(right_bytes):
        return False
    return secrets.compare_digest(left_bytes, right_bytes)
